plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c

```diff
@@ -0,0 +1,374 @@
+	os_event_destroy(stream->stop_event);
+	bfree(stream);
+	return NULL;
+}
+
+static int process_packet(struct ffmpeg_muxer *stream)
+{
+	struct encoder_packet *packet;
+	int remaining;
+
+	pthread_mutex_lock(&stream->write_mutex);
+	remaining = stream->mux_packets.num;
+	if (remaining) {
+		packet = &stream->mux_packets.array[0];
+		write_packet(stream, packet);
+		obs_encoder_packet_release(packet);
+		da_erase(stream->mux_packets, 0);
+	}
+	pthread_mutex_unlock(&stream->write_mutex);
+
+	return 0;
+}
+
+static void *write_thread(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+	stream->active = true;
+
+	while (os_sem_wait(stream->write_sem) == 0) {
+
+		if (os_event_try(stream->stop_event) == 0)
+			break;
+
+		int ret = process_packet(stream);
+		if (ret != 0) {
+			int code = OBS_OUTPUT_ERROR;
+
+			if (ret == -ENOSPC)
+				code = OBS_OUTPUT_NO_SPACE;
+
+			obs_output_signal_stop(stream->output, code);
+			hls_deactivate(stream, 0);
+			break;
+		}
+	}
+
+	stream->active = false;
+	return NULL;
+}
+
+bool ffmpeg_hls_mux_start(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+	obs_service_t *service;
+	const char *path_str;
+	const char *stream_key;
+	struct dstr path = {0};
+	obs_encoder_t *vencoder;
+	obs_data_t *settings;
+	int keyint_sec;
+
+	if (!obs_output_can_begin_data_capture(stream->output, 0))
+		return false;
+	if (!obs_output_initialize_encoders(stream->output, 0))
+		return false;
+
+	service = obs_output_get_service(stream->output);
+	if (!service)
+		return false;
+	path_str = obs_service_get_url(service);
+	stream_key = obs_service_get_key(service);
+	dstr_copy(&stream->stream_key, stream_key);
+	dstr_copy(&path, path_str);
+	dstr_replace(&path, "{stream_key}", stream_key);
+	dstr_init(&stream->muxer_settings);
+	dstr_catf(&stream->muxer_settings,
+		  "http_user_agent=libobs/%s method=PUT http_persistent=1",
+		  OBS_VERSION);
+	dstr_catf(&stream->muxer_settings, " ignore_io_errors=1");
+
+	vencoder = obs_output_get_video_encoder(stream->output);
+	settings = obs_encoder_get_settings(vencoder);
+	keyint_sec = obs_data_get_int(settings, "keyint_sec");
+	if (keyint_sec)
+		dstr_catf(&stream->muxer_settings, " hls_time=%d", keyint_sec);
+
+	obs_data_release(settings);
+
+	start_pipe(stream, path.array);
+	dstr_free(&path);
+
+	if (!stream->pipe) {
+		obs_output_set_last_error(
+			stream->output, obs_module_text("HelperProcessFailed"));
+		warn("Failed to create process pipe");
+		return false;
+	}
+
+	/* write headers and start capture */
+	os_atomic_set_bool(&stream->active, true);
+	os_atomic_set_bool(&stream->capturing, true);
+	os_atomic_set_bool(&stream->threading_buffer, true);
+	stream->total_bytes = 0;
+	stream->dropped_frames = 0;
+	stream->min_priority = 0;
+
+	obs_output_begin_data_capture(stream->output, 0);
+
+	dstr_copy(&stream->printable_path, path_str);
+	info("Writing to path '%s'...", stream->printable_path.array);
+
+	stream->mux_thread_joinable = pthread_create(&stream->mux_thread, NULL,
```

Comment:
Move this above os_atomic_set_bool(&stream->active, true); Return false if pthread_create failed.